sdk/storage/Azure.Storage.Internal.Avro/src/AvroReader.cs

```diff
@@ -0,0 +1,196 @@
+        /// in the middle of an Avro file.
+        /// </summary>
+        public AvroReader(
+            Stream dataStream,
+            Stream headerStream,
+            long currentBlockOffset,
+            long indexWithinCurrentBlock)
+        {
+            _dataStream = dataStream;
+            _headerStream = headerStream;
+            _metadata = new Dictionary<string, string>();
+            _initalized = false;
+            BlockOffset = currentBlockOffset;
+            ObjectIndex = indexWithinCurrentBlock;
+            _initalized = false;
+        }
+
+        private async Task Initalize(bool async, CancellationToken cancellationToken = default)
+        {
+            // Four bytes, ASCII 'O', 'b', 'j', followed by 1.
+            byte[] header = await AvroParser.ReadFixedBytesAsync(_headerStream, AvroConstants.InitBytes.Length, async, cancellationToken).ConfigureAwait(false);
+            if (!header.SequenceEqual(AvroConstants.InitBytes))
+            {
+                throw new ArgumentException("Stream is not an Avro file.");
+            }
+
+            // File metadata is written as if defined by the following map schema:
+            // { "type": "map", "values": "bytes"}
+            _metadata = await AvroParser.ReadMapAsync(_headerStream, AvroParser.ReadStringAsync, async, cancellationToken).ConfigureAwait(false);
+
+            // Validate codec
+            _metadata.TryGetValue(AvroConstants.CodecKey, out string codec);
+            if (!(codec == null || codec == "null"))
+            {
+                throw new ArgumentException("Codecs are not supported");
+            }
+
+            // The 16-byte, randomly-generated sync marker for this file.
+            _syncMarker = await AvroParser.ReadFixedBytesAsync(_headerStream, AvroConstants.SyncMarkerSize, async, cancellationToken).ConfigureAwait(false);
+
+            // Parse the schema
+            using JsonDocument schema = JsonDocument.Parse(_metadata[AvroConstants.SchemaKey]);
+            _itemType = AvroType.FromSchema(schema.RootElement);
+
+            if (BlockOffset == 0)
+            {
+                BlockOffset = _dataStream.Position;
+            }
+
+            // Populate _itemsRemainingInCurrentBlock
+            _itemsRemainingInBlock = await AvroParser.ReadLongAsync(_dataStream, async, cancellationToken).ConfigureAwait(false);
+
+            // skip block length
+            await AvroParser.ReadLongAsync(_dataStream, async, cancellationToken).ConfigureAwait(false);
+
+            _initalized = true;
+
+            if (ObjectIndex > 0)
+            {
+                for (int i = 0; i < ObjectIndex; i++)
+                {
+                    await _itemType.ReadAsync(_dataStream, async, cancellationToken).ConfigureAwait(false);
+                    _itemsRemainingInBlock--;
+                }
+            }
+        }
+
+        public bool HasNext() => !_initalized || _itemsRemainingInBlock > 0;
```

Comment:
Is it better we call `Initalize()` here in case the avro file contains 0 record?